distogram/__init__.py

```diff
@@ -1,240 +1,203 @@
     _update_diffs(h, i)
     return h
 
 
-def _compute_diffs(h):
-    diffs = []
-    bins = h.bins
-    h.min_diff = None
-    for index in range(1, len(bins)):
-        diff = _weighted_diff(h, h.bins[index], h.bins[index-1])
-        diffs.append(diff)
-        if h.min_diff is None or diff < h.min_diff:
-            h.min_diff = diff
+def _compute_diffs(h: Distogram) -> List[float]:
+    diffs = [
+        _weighted_diff(h, b2, b1)
+        for b1, b2 in zip(h.bins[:-1], h.bins[1:])
+    ]
+    h.min_diff = min(diffs)
 
     return diffs
 
 
-def _linear_index(bins, value):
-    ratio = float((value - bins[0][0]) / (bins[-1][0] - bins[0][0]))
-    index = round(ratio * len(bins))
-    if index > 0 and index < (len(bins) - 1):
-        if value > bins[index][0] and value < bins[index+1][0]:
-            return index+1
-        if value > bins[index-1][0] and value < bins[index][0]:
-            return index
-
-    return 0
-
-
-def _search_in_place_index(h, new_value, index):
-    bins = h.bins
-
+def _search_in_place_index(h: Distogram, new_value: float, index: int) -> int:
     if h.diffs is None:
         h.diffs = _compute_diffs(h)
 
-    min_diff = h.min_diff
-    i_bin = None
-
     if index > 0:
-        diff1 = _weighted_diff(h, (new_value, 1), bins[index-1])
-        diff2 = _weighted_diff(h, bins[index], (new_value, 1))
-        if diff1 < diff2:
-            diff = diff1
-            i_bin = index-1
-        else:
-            diff = diff2
-            i_bin = index
+        diff1 = _weighted_diff(h, (new_value, 1), h.bins[index - 1])
+        diff2 = _weighted_diff(h, h.bins[index], (new_value, 1))
+
+        i_bin, diff = (index - 1, diff1) if diff1 < diff2 else (index, diff2)
+
+        return i_bin if diff < h.min_diff else -1
 
-        if diff < min_diff:
-            return i_bin
-        else:
-            return -1
     return -1
 
 
-def update(h, value, count=1):
-    '''Adds a new element to the distribution.
+def update(h: Distogram, value: float, c: int = 1) -> Distogram:
+    """ Adds a new element to the distribution.
 
     Args:
         h: A Distogram object.
         value: The value to add on the histogram.
-        count: [Optional] The number of times that value must be added.
+        c: [Optional] The number of times that value must be added.
 
     Returns:
         A Distogram object where value as been processed.
-    '''
-    bins = h.bins
+    """
     index = 0
-    if len(bins) > 0:
-        if value <= bins[0][0]:
+    if len(h.bins) > 0:
+        v0, _ = h.bins[0]
+        vl, _ = h.bins[-1]
+
+        if value <= v0:
             index = 0
-        elif value >= bins[-1][0]:
+        elif value >= vl:
             index = -1
         else:
-            if len(bins) >= h.bin_count:
-                index = _linear_index(bins, value)
-            if index == 0:
-                index = bisect_left(bins, (value, 1))
+            index = bisect_left(h.bins, (value, 1))
```

Comment:
This passes all the unit tests. I did not change, add or remove any tests so if I'm wrong on this, please let me know.

Reply:
I checked and made some tests: _linear_index was an optimization before I use the stdlib implementation of bisect_left. Using only bisect_left is now much faster than the linear interpolation. So we must drop _linear_index.